metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/usage.py

```diff
@@ -158,113 +301,147 @@ def __init__(self, config: BigQueryV2Config, report: BigQueryV2Report):
+
+    def _ingest_events(
+        self,
+        projects: Iterable[str],
+        table_refs: Set[str],
+        usage_state: BigQueryUsageState,
+    ) -> None:
+        """Read log and store events in usage_state."""
+        num_aggregated = 0
+        for project in projects:
+            self.report.set_project_state(project, "Usage Extraction Ingestion")
+            for audit_event in self._get_usage_events(project):
+                try:
+                    num_aggregated += self._store_usage_event(
+                        audit_event, usage_state, table_refs
                     )
-                else:
-                    parsed_events = self._parse_bigquery_log_entries(
-                        bigquery_log_entries
+                except Exception:
+                    logger.warning(
+                        f"Unable to store usage event {audit_event}", exc_info=True
                     )
+        logger.info(f"Total number of events aggregated = {num_aggregated}.")
 
-                hydrated_read_events = self._join_events_by_job_id(parsed_events)
-                # storing it all in one big object.
-
-                # TODO: handle partitioned tables
-
-                # TODO: perhaps we need to continuously prune this, rather than
-                num_aggregated: int = 0
-                self.report.num_operational_stats_workunits_emitted = 0
-                for event in hydrated_read_events:
-                    if self.config.usage.include_operational_stats:
-                        operational_wu = self._create_operation_aspect_work_unit(event)
-                        if operational_wu:
-                            yield operational_wu
-                            self.report.num_operational_stats_workunits_emitted += 1
-                    if event.read_event:
-                        self._aggregate_enriched_read_events(
-                            aggregated_info, event, tables
-                        )
-                        num_aggregated += 1
-                logger.info(f"Total number of events aggregated = {num_aggregated}.")
-                bucket_level_stats: str = "\n\t" + "\n\t".join(
-                    [
-                        f'bucket:{db.strftime("%m-%d-%Y:%H:%M:%S")}, size={len(ads)}'
-                        for db, ads in aggregated_info.items()
-                    ]
+    def _generate_operational_workunits(
+        self, usage_state: BigQueryUsageState
+    ) -> Iterable[MetadataWorkUnit]:
+        self.report.set_project_state("All", "Usage Extraction Operational Stats")
+        for audit_event in usage_state.standalone_events():
+            try:
+                operational_wu = self._create_operation_workunit(audit_event)
+                if operational_wu:
+                    yield operational_wu
+                    self.report.num_operational_stats_workunits_emitted += 1
+            except Exception:
+                logger.warning(
```

Comment:
Same thing here, trying to be "defensive" when writing the sources so that our ingestion runs are more robust. Trying to protect against not only potential failures from expected exceptions or unreliable 3rd party calls, but our own code too

Reply:
Now handled purely by `report.num_usage_failures` dict